pkg/logs/decoder/auto_multiline_handler.go

```diff
@@ -0,0 +1,221 @@
+	d.Lock()
+	defer d.Unlock()
+	return d.pattern
+}
+
+// AutoMultilineHandler can attempts to detect a known/commob pattern (a timestamp) in the logs
+// and will switch to a MultiLine handler if one is detected and the thresholds are met.
+type AutoMultilineHandler struct {
+	multiLineHandler  *MultiLineHandler
+	singleLineHandler *SingleLineHandler
+	inputChan         chan *Message
+	outputChan        chan *Message
+	isRunning         bool
+	linesToAssess     int
+	linesTested       int
+	lineLimit         int
+	matchThreshold    float64
+	scoredMatches     []*scoredPattern
+	processsingFunc   func(message *Message)
+	flushTimeout      time.Duration
+	source            *config.LogSource
+	timeoutTimer      *time.Timer
+	detectedPattern   *DetectedPattern
+}
+
+// NewAutoMultilineHandler returns a new AutoMultilineHandler.
+func NewAutoMultilineHandler(outputChan chan *Message,
+	lineLimit, linesToAssess int,
+	matchThreshold float64,
+	matchTimeout time.Duration,
+	flushTimeout time.Duration,
+	source *config.LogSource,
+	additionalPatterns []*regexp.Regexp,
+	detectedPattern *DetectedPattern,
+) *AutoMultilineHandler {
+
+	// Put the user patterns at the beginning of the list so we prioritize them if there is a conflicting match.
+	patterns := append(additionalPatterns, formatsToTry...)
+
+	scoredMatches := make([]*scoredPattern, len(patterns))
+	for i, v := range patterns {
+		scoredMatches[i] = &scoredPattern{
+			score:  0,
+			regexp: v,
+		}
+	}
+	h := &AutoMultilineHandler{
+		inputChan:       make(chan *Message),
+		outputChan:      outputChan,
+		isRunning:       true,
+		lineLimit:       lineLimit,
+		matchThreshold:  matchThreshold,
+		scoredMatches:   scoredMatches,
+		linesToAssess:   linesToAssess,
+		flushTimeout:    flushTimeout,
+		source:          source,
+		timeoutTimer:    time.NewTimer(matchTimeout),
+		detectedPattern: detectedPattern,
+	}
+
+	h.singleLineHandler = NewSingleLineHandler(outputChan, lineLimit)
+	h.processsingFunc = h.processAndTry
+
+	return h
+}
+
+// Handle puts all new lines into a channel for later processing.
+func (h *AutoMultilineHandler) Handle(input *Message) {
+	h.inputChan <- input
+}
+
+// Stop stops the handler.
+func (h *AutoMultilineHandler) Stop() {
+	close(h.inputChan)
+}
+
+// Start starts the handler.
+func (h *AutoMultilineHandler) Start() {
+	go h.run()
+}
+
+// run consumes new lines and processes them.
+func (h *AutoMultilineHandler) run() {
+	for {
+		if !h.isRunning {
+			return
+		}
+		line, isOpen := <-h.inputChan
+		if !isOpen {
+			close(h.outputChan)
+			return
+		}
+		h.processsingFunc(line)
+	}
+}
+
+func (h *AutoMultilineHandler) processAndTry(message *Message) {
+	// Process message before anything else
+	h.singleLineHandler.process(message)
+
+	for i, scoredPattern := range h.scoredMatches {
+		match := scoredPattern.regexp.Match(message.Content)
+		if match {
+			log.Tracef("A regexp matched during multiline auto sensing: %v", scoredPattern.regexp)
```

Comment:
It's `Trace` so we are fine for regular usage but, wouldn't this be flooding the output log if a file is filled with a pattern? I'm just wondering if we should not log a trace (or debug) when leaving this for if `scoredPattern.score` has changed?
I may be overcautious here, please ignore if I'm exaggerating.

Reply:
you are right. ill remove this log line - there are other useful log lines later on for the cumulative match results.